[REVIEW] Initial work for decimal type in Java/JNI [skip ci]

nartal1 opened the pull request:
This partially addresses tasks from  issue https://github.com/rapidsai/cudf/issues/6515.
The goal of this PR is to add DType class and make the necessary changes in Java/JNI code to support scale in DType. 

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

revans2 commented on java/src/main/java/ai/rapids/cudf/DataType.java:
```diff
@@ -0,0 +1,52 @@
+
+public class DataType {
+
+  DType typeId;
```
Generally we have had `ColumnVector` be an immutable. I would prefer to have DataType be similar.  Once it is created I don't see a lot of value in changing it.

nartal1 replied:
Thanks! Will update this in next commit.

revans2 commented on java/src/main/java/ai/rapids/cudf/DataType.java:
```diff
@@ -0,0 +1,52 @@
+  int scale;
+   public DataType(DType id) {typeId = id; }
+
+   public DataType(DType id,  int fp_scale) {
```
Generally we use `camelCase` instead of `snake_case` for variable names in java.

nartal1 replied:
Sorry, missed this. Was switching from cpp.

revans2 reviewed (commented):
I know this is still WIP, but I wanted to call out a few things.

nartal1 commented on java/src/main/java/ai/rapids/cudf/DataType.java:
```diff
@@ -0,0 +1,52 @@
+   }
+  private static native long makeNativeType(int typeId, int scale);
+
+  public class NativeDataType implements AutoCloseable {
```
Thanks for pointing this out. I was contemplating on this. 
Right now, we don't have any use-case, we could pass DType ordinal and scale(as I have done in current implementation) but if we need to add another field such as `precision` in DataType later, it would be easier to add.  
I will go ahead with current approach for now i.e pass ordinal and the scale.

sperlingxx replied:
Hi @jlowe , as previous discussion between @nartal1 and I, we decide to introduce the class `NativeDataType`, which holds/manages the handle of native class instance `data_type`, because it can pass the handle to JNI method instead of typeID, scale and possible additional parameters in future. In this approach, we can avoid interface change for any changes on `DataType`.

class `NativeDataType` needs to be `AutoClosable` in case of memory leak in native level.  It should be initialized with try resource clause.

nartal1 replied:
Hi @sperlingxx, Thanks for reviewing. Had a discussion with @jlowe on this today morning. It is correct that we can avoid interface change for any changes on DataType in the future but it adds complexity of tracking the lifetimes(which may be prone to bugs if not closed properly).  As of now we are passing one more arg i.e `scale` in the JNI api's which is okay. As of now, there doesn't seem to be much added in `data_type` itseems and we can revisit this approach in the future if it becomes absolutely necessary.

sperlingxx replied:
Okay, I believe you are right.

sperlingxx commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -3729,6 +3767,14 @@ public static ColumnVector fromDoubles(double... values) {
     return build(DType.FLOAT64, values.length, (b) -> b.appendArray(values));
   }
 
+ /* public static ColumnVector fromDecimal32(int... values) {
```
involve this commented codes later till we are fully ready?

sperlingxx reviewed (commented):
LGTM in general! I think we can merge this PR for fast iteration. 

As we discussed, we will introduce a new class `NativeDataType` in next PR, which holds/manages the handle of native class instance `data_type`. In this approach, it can pass the handle to JNI method instead of typeID, scale and possible additional parameters in future. 

class `NativeDataType` needs to be `AutoClosable` in case of memory leak in native level.  It should be initialized with try resource clause.

nartal1 commented:
I think the code in `Builder` and `ColumnBuilder` also needs to be changed from DType to DataType.

revans2 commented:
> I think the code in `Builder` and `ColumnBuilder` also needs to be changed from DType to DataType.

If it is simple to do yes, but if it is not simple we should look at only supporting the newer one that supports nested types and deprecating the older one.  No point in maintaing two APIs that do the same thing.

nartal1 commented:
@revans2 @jlowe It would be great if you could take a look and let me know if this is something in the right direction. As of now I am making changes and making sure we don't break any current tests. The refactor is in different places so would be helpful if I could get early feedback  on whether to continue or if we need to pivot.

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -50,6 +50,11 @@
     this(type, rows, nullCount, hostDataBuffer, hostValidityBuffer, null);
   }
 
+  HostColumnVector(ai.rapids.cudf.DataType type, long rows, Optional<Long> nullCount,
```
In HostColumnVector there is an abstract class `DataType`. So I am using fully qualified name in this file. Please let me know if you prefer any other name to represent DataType class(with DType + scale) if this is confusing. I wasn't sure what else would represent `DataType` : ) 
https://github.com/rapidsai/cudf/blob/branch-0.17/java/src/main/java/ai/rapids/cudf/HostColumnVector.java#L1545

kuhushukla replied:
You don't need the fully qualified name imo. That's what Jason is saying I think

jlowe replied:
Ah, sorry I totally missed that `HostColumnVector` already has a `DataType`.  It would be ideal to not collide with this name, but I'm not coming up with a better alternative name offhand.

nartal1 replied:
Thanks! Will keep it as it is for now and update it later if/when we come up with better name.

kuhushukla replied:
I'm sorry I haven't really looked into this change, but why can't we change and tweak the one that exists?

razajafri replied:
I agree with @kuhushukla can we combine the two? 

```
class DataType {
   DType typeId
   int scale 
   def isNullable()
   def getSize() // may be return size in bytes for primitive types
}

class NestedType extends DataType {
   def getNumChildren()
   def getChild(int index) 
}

class StructType extends NestedType
class ListType extends NestedType

```

revans2 replied:
I agree that we need to figure out long term how we want to handle nested schemas.  Currently we have a Schema class that is used with CSV parsing, DType and DataType.  I think this needs to be something where someone gathers all of the requirements and comes up with a design for what we really want this to look like.  I don't think that has been done yet.

revans2 replied:
I also want to add that part of the reason we did it this way was to avoid most issues with source incompatibility. If we do what is being proposed haphazardly we will break anyone using the APIs so we need to be careful.  The current changes should be very close to source compatible.  Minor changes if any.  We can get away with some of that because we have singleton objects.

The other issue I want to call out is that cudf itself has no schema equivalent right now.  So if we do combine them all together then we will have no clear mapping between this java object and a corresponding C++ object.  I am not sure that is ideal.

razajafri commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -3088,7 +3114,7 @@ public long getColumnViewAddress() {
   @Override
   public ColumnViewAccess getChildColumnViewAccess(int childIndex) {
-    if (!type.isNestedType()) {
+    if (!type.typeId.isNestedType()) {
```
I would parrot @jlowe 's comment here to add the `isNestedType` to DataType to reduce code changes

nartal1 replied:
Done

razajafri commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -1239,48 +1281,48 @@ public Builder appendUTF8String(byte[] value, int offset, int length) {
       assert (values.length + currentIndex) <= rows;
-      assert type == DType.INT8 || type == DType.UINT8 || type == DType.BOOL8;
-      data.setBytes(currentIndex * type.sizeInBytes, values, 0, values.length);
+      assert type.typeId == DType.INT8 || type.typeId == DType.UINT8 || type.typeId == DType.BOOL8;
```
Now that we are making a change this would be a good time to clean this up and have a helper method in DataType like the one we have e.g. 'isTimestamp` etc

nartal1 replied:
Done

razajafri commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -1239,48 +1281,48 @@ public Builder appendUTF8String(byte[] value, int offset, int length) {
 
     public Builder appendArray(short... values) {
-      assert type == DType.INT16 || type == DType.UINT16;
+      assert type.typeId == DType.INT16 || type.typeId == DType.UINT16;
```
same comment here. can we have a helper?

nartal1 replied:
Done

razajafri commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -273,28 +286,28 @@ private void assertsForGet(long index) {
    */
   public final short getShort(long index) {
-    assert type == DType.INT16 || type == DType.UINT16 : type + " is not stored as a short.";
+    assert type.typeId == DType.INT16 || type.typeId == DType.UINT16 : type + " is not stored as a short.";
```
same comment here. have a helper

nartal1 replied:
Done

razajafri commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -273,28 +286,28 @@ private void assertsForGet(long index) {
    */
   public byte getByte(long index) {
-    assert type == DType.INT8 || type == DType.UINT8 || type == DType.BOOL8 : type +
+    assert type.typeId == DType.INT8 || type.typeId == DType.UINT8 || type.typeId == DType.BOOL8 : type +
```
same comment here. have a helper

nartal1 replied:
Done

razajafri commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1364,6 +1369,18 @@ JNIEXPORT jint JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeTypeId(JNIEnv *
+  try {
+    cudf::jni::auto_set_device(env);
+    cudf::column_view *column = reinterpret_cast<cudf::column_view *>(handle);
+    return static_cast<jint>(column->type().scale());
```
is type checking needed before calling `scale()` on  type_id? will it blow up in case its not a decimal type?

revans2 replied:
We should not need the cast and I would prefer to not have it because it shows that we don't have issues mapping from the native scale to the JNI representation.

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -165,18 +185,18 @@ synchronized int getRefCount() {
     DeviceMemoryBuffer valid = null;
     DeviceMemoryBuffer offsets = null;
     try {
-      if (!type.isNestedType()) {
```
Done.

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -165,18 +185,18 @@ synchronized int getRefCount() {
   public ColumnVector copyToDevice() {
     if (rows == 0) {
-      return new ColumnVector(type, 0, Optional.of(0L), null, null, null);
+      return new ColumnVector(type.typeId, 0, Optional.of(0L), null, null, null);
```
Updated it. Have to still look into Nested types more.

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -1119,72 +1161,72 @@ public String toString() {
 
     public final Builder append(int value) {
-      assert type.isBackedByInt();
+      assert type.typeId.isBackedByInt();
```
Done

nartal1 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1364,6 +1369,18 @@ JNIEXPORT jint JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeTypeId(JNIEnv *
 }
 
+JNIEXPORT jint JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeTypeScale(JNIEnv *env,
+                                                                        jobject j_object,
```
Done

nartal1 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -145,9 +145,13 @@ public ColumnVector(long nativePointer) {
-    this.type = offHeap.getNativeType();
     this.rows = offHeap.getNativeRowCount();
 
+    if (offHeap.getNativeType() == DType.DECIMAL32 || offHeap.getNativeType() == DType.DECIMAL64) {
```
Thanks for pointing this out. Done.

nartal1 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -1053,7 +1076,7 @@ public ColumnVector month() {
    */
   public ColumnVector day() {
-    assert type.isTimestamp();
+    assert type.typeId.isTimestamp();
```
Done

nartal1 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -1065,7 +1088,7 @@ public ColumnVector day() {
    */
   public ColumnVector hour() {
-    assert type.hasTimeResolution();
+    assert type.typeId.hasTimeResolution();
```
Done

nartal1 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -1360,7 +1383,7 @@ static long binaryOp(ColumnVector lhs, Scalar rhs, BinaryOp op, DType outputType
    */
   public Scalar sum() {
-    return sum(type);
+    return sum(type.typeId);
```
You are right, In this iteration i was planning  to fix the compilation error and thought of changing it while adding fixed-point support.

jlowe replied:
The problem with that approach is it creates a bunch of code changes that need to be reviewed in this PR and then a followup PR that then removes all of those changes which again needs to be reviewed.  I suggest adding the `sum(DataType)` and similar methods in this PR (they need to be added at some point anyway).  Have them assert/throw on decimal types if we're not ready to do decimals yet in this PR.  That's less overall code churn and stuff to review and is a more direct line to what we need in the end.

nartal1 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -63,6 +63,8 @@
   DURATION_NANOSECONDS(8, 21, "int64"),
   //DICTIONARY32(4, 22, "NO IDEA"),
 
+  DECIMAL32(4, 25, "decimal32"),
```
Sorry,  My bad. Missed it in hurry. Corrected it.

jlowe commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -271,7 +271,7 @@ private static void assertStructColumnsEquals(HostColumnVectorCore expected, Hos
     boolean hasNestedTypeChildren = false;
     for (HostColumnVectorCore expectedChild : expected.getNestedChildren()) {
-      if (expectedChild.type.isNestedType()) {
+      if (expectedChild.type.typeId.isNestedType()) {
```
Same suggestion of adding this predicate as a delegating method to `DataType`.

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -141,9 +141,14 @@ public ColumnVector(long nativePointer) {
     this.rows = offHeap.getNativeRowCount();
+    DType nativeType = offHeap.getNativeType();
 
+    if (nativeType == DType.DECIMAL32 || nativeType == DType.DECIMAL64) {
```
I suggest adding an `EnumSet` for decimal types and a predicate method like `isDecimal` so it easier and cheaper to check for a decimal type.  This is already done for other sets of `DType` values.

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -195,11 +216,37 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
     for (int i = 0; i < nestedColumnVectors.size(); i++) {
       childHandles[i] = nestedColumnVectors.get(i).getViewHandle();
     }
-    offHeap = new OffHeapState(type, (int) rows, nullCount, dataBuffer, validityBuffer, offsetBuffer,
```
This, and every other constructor/method, that takes a `DType` rather than a `DataType` should assert/throw if it's a decimal type since that's an incomplete type.

Using something like the suggested `toDataType` method and implementing all existing `DType` methods in terms of their `DataType` versions would cover that.  We probably should also deprecate all the existing `DType` methods, pointing to the `DataType` version instead.

revans2 replied:
I agree and we should look at each API and consider deprecating the versions that only take a DType.

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -195,11 +216,37 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
     for (int i = 0; i < nestedColumnVectors.size(); i++) {
```
This constructor should be implemented in terms of the new `DataType` constructor to minimize the duplication of code and reduce maintenance.  For example, something like this:
```java
private static DataType toDataType(DType type) {
  assert !type.isDecimal() : "Cannot convert a decimal type without a specified scale";
  return new DataType(type);
}

[...]

public ColumnVector(DType type, long rows, Optional<Long> nullCount,
                    DeviceMemoryBuffer dataBuffer, DeviceMemoryBuffer validityBuffer,
                    DeviceMemoryBuffer offsetBuffer, List<NestedColumnVector> nestedColumnVectors) {
  this(toDataType(type), rows, nullCount, dataBuffer, validityBuffer, offsetBuffer, nestedColumnVectors);
}
```

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -170,6 +175,22 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
       assert offsetBuffer == null : "offsets are only supported for STRING";
     }
 
+    long[] children = new long[] {};
```
This constructor should be implemented in terms of the new `DataType` constructor version.

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -217,10 +264,11 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
     offHeap = new OffHeapState(viewAddress, contiguousBuffer);
     MemoryCleaner.register(this, offHeap);
-    this.type = offHeap.getNativeType();
+    this.type = new DataType(offHeap.getNativeType());
```
What about a possible scale here if the type is a decimal type?

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -450,7 +498,7 @@ public HostColumnVector copyToHost() {
       BaseDeviceMemoryBuffer offsets = getOffsetBuffer();
       BaseDeviceMemoryBuffer data = null;
-      DType type = this.type;
+      DType type = this.type.typeId;
```
Shouldn't this be using `this.type` here?  Once the `isNestedType` predicate is added, we will need to pass a `DataType` to the `HostColumnVector` constructor so it can access the scale to properly decode decimal types.

jlowe commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1620,12 +1637,18 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeColumnView(JNI
 
 JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn(JNIEnv *env,
                                                                              jobject j_object,
```
```suggestion
                                                                             jclass,
```

jlowe commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1364,6 +1369,18 @@ JNIEXPORT jint JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeTypeId(JNIEnv *
+}
+
 JNIEXPORT jint JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeRowCount(JNIEnv *env,
                                                                           jobject j_object,
```
```suggestion
                                                                          jclass,
```

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -170,6 +175,22 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
+  public ColumnVector(DataType type, long rows, Optional<Long> nullCount,
+                      DeviceMemoryBuffer dataBuffer, DeviceMemoryBuffer validityBuffer,
+                      DeviceMemoryBuffer offsetBuffer) {
+
```
The String assert here from the other constructor is lost.  Is that expected?

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -170,6 +175,22 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
     }
 
+    long[] children = new long[] {};
+    DataType dataType = new DataType(type);
```
I would almost rather have us throw an exception if you try to use this constructor with a decimal type, and deprecate this constructor.  It could cause a lot of subtle bugs if we try to use it with a Decimal type and the scale is lost.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -131,7 +144,7 @@ public long getNullCount() {
    */
   @Override
   public DType getDataType() {
```
This is one case where we might want to break backwards compatibility and just drop DType everywhere and replace it with DataType across the board.

jlowe commented:
> and also public static final instances for each non DECIMAL type

I was already thinking along these lines, but I hadn't considered taking over `DType` itself.  I'm generally +1 on the idea of commandeering `DType`, although we will have a "knife switch" event when this gets checked in.  The snapshot build published afterwards is going to break anything that hasn't recompiled.

I'm still worried about all the places code assumes checking for equality on a data type can be performed with `==` which is no longer the case, but we need to worry about that whether it's called `DataType` or `DType`.  Not calling it `DataType` nicely side-steps the problem identified in `HostColumnVector` and also the upcoming name confusion with Spark's `DataType`.

nartal1 commented:
Thanks @revans2 and @jlowe for reviewing and your inputs on new design. I will rework on this and update it here. Liked this idea where it solves confusion on naming `DataType`.

sperlingxx commented:
Hi @revans2 @jlowe ,  @nartal1 and I are trying to implement the idea `DType with DTypeEnum`. Both of us want to ensure whether we fully understand this idea. Here are some code snippets to elaborate our understanding:
 
```java
  public static final DType EMPTY = new DType(DTypeEnum.EMPTY);
  public static final DType INT8 = new DType(DTypeEnum.INT8);
  public static final DType INT16 = new DType(DTypeEnum.INT16);
  public static final DType INT32 = new DType(DTypeEnum.INT32);
  public static final DType INT64 = new DType(DTypeEnum.INT64);
  public static final DType UINT8 = new DType(DTypeEnum.UINT8);
  public static final DType UINT16 = new DType(DTypeEnum.UINT16);
  public static final DType UINT32 = new DType(DTypeEnum.UINT32);
  public static final DType UINT64 = new DType(DTypeEnum.UINT64);
  public static final DType FLOAT32 = new DType(DTypeEnum.FLOAT32);
  public static final DType FLOAT64 = new DType(DTypeEnum.FLOAT64);
  public static final DType BOOL8 = new DType(DTypeEnum.BOOL8);
  public static final DType TIMESTAMP_DAYS = new DType(DTypeEnum.TIMESTAMP_DAYS);
  public static final DType TIMESTAMP_SECONDS = new DType(DTypeEnum.TIMESTAMP_SECONDS);
  public static final DType TIMESTAMP_MILLISECONDS = new DType(DTypeEnum.TIMESTAMP_MILLISECONDS);
  public static final DType TIMESTAMP_MICROSECONDS = new DType(DTypeEnum.TIMESTAMP_MICROSECONDS);
  public static final DType TIMESTAMP_NANOSECONDS = new DType(DTypeEnum.TIMESTAMP_NANOSECONDS);
  public static final DType DURATION_DAYS = new DType(DTypeEnum.DURATION_DAYS);
  public static final DType DURATION_SECONDS = new DType(DTypeEnum.DURATION_SECONDS);
  public static final DType DURATION_MILLISECONDS = new DType(DTypeEnum.DURATION_MILLISECONDS);
  public static final DType DURATION_MICROSECONDS = new DType(DTypeEnum.DURATION_MICROSECONDS);
  public static final DType DURATION_NANOSECONDS = new DType(DTypeEnum.DURATION_NANOSECONDS);
  public static final DType STRING = new DType(DTypeEnum.STRING);
  public static final DType LIST = new DType(DTypeEnum.LIST);
  public static final DType STRUCT = new DType(DTypeEnum.STRUCT);

  private static final DType[] staticDTypes = new DType[] {
      EMPTY,
      INT8,
      INT16,
      INT32,
      INT64,
      UINT8,
      UINT16,
      UINT32,
      UINT64,
      FLOAT32,
      FLOAT64,
      BOOL8,
      TIMESTAMP_DAYS,
      TIMESTAMP_SECONDS,
      TIMESTAMP_MILLISECONDS,
      TIMESTAMP_NANOSECONDS,
      DURATION_DAYS,
      DURATION_SECONDS,
      DURATION_MILLISECONDS,
      DURATION_MICROSECONDS,
      DURATION_NANOSECONDS,
      null, // DICTIONARY32
      STRING,
      LIST,
      null, // DECIMAL32
      null, // DECIMAL64
      STRUCT
  };

  public static DType fromNative(int nativeId, int scale) {
    if (nativeId >=0 && nativeId < staticDTypes.length) {
      if (staticDTypes[nativeId] != null) {
        return staticDTypes[nativeId];
      }
      if (nativeId == DTypeEnum.DECIMAL32.nativeId) {
        // TODO: maybe we can verify scale at here?
        return new DType(DTypeEnum.DECIMAL32, scale);
      }
      if (nativeId == DTypeEnum.DECIMAL64.nativeId) {
        // TODO: maybe we can verify scale at here?
        return new DType(DTypeEnum.DECIMAL64, scale);
      }
    }
    throw new IllegalArgumentException("Could not translate " + nativeId + " into a DType");
  }

  public static DType fromNative(int nativeId) {
    // TODO: replace Integer.MIN_VALUE with a constant as noop placeholder ?
    return fromNative(nativeId, Integer.MIN_VALUE);
  }
```

nartal1 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -196,6 +215,17 @@ public static DType fromNative(int nativeId) {
+  );
+
+  private static final EnumSet<DType> BYTES = EnumSet.of(
+          DType.INT8,
```
done

nartal1 commented on java/src/main/java/ai/rapids/cudf/DataType.java:
```diff
@@ -0,0 +1,46 @@
+public class DataType {
+
+  final DType typeId;
+  int scale = 0;
```
done

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -175,8 +195,8 @@ public ColumnVector copyToDevice() {
         if (hdata != null) {
-          long dataLen = rows * type.sizeInBytes;
-          if (type == DType.STRING) {
+          long dataLen = rows * type.typeId.sizeInBytes;
```
done

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+  public static DType fromNative(int nativeId, int scale) {
+    if (nativeId >=0 && nativeId < staticDTypes.length) {
+      if (staticDTypes[nativeId] != null) {
+        return staticDTypes[nativeId];
```
I'm guessing @revans2 would like the original logic preserved where it would scan the types checking for the first one that matched native IDs.  At a minimum it needs to verify that the native ID of the DType enum it loaded matches the one specified and throw if it doesn't (meaning there's a programming error on the order of the table).  I think it's very likely someone will accidentally update one of these tables without the other (e.g.: add a new DType enum value and forget to put a `null` in the corresponding singleton array.  Comments above each of these lists stating it's important to keep the other list in sync would help.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+  }
+
+  public static DType fromNative(int nativeId) {
+    return fromNative(nativeId, Integer.MIN_VALUE);
```
IMO we should not be implicitly setting the scale of decimal types.  Trying to convert from a native ID to a decimal type without specifying a scale should be an error.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
    */
-  TIMESTAMP_SECONDS(8, 13, "timestamp[s]"),
+  public DType(DTypeEnum id, int decimalScale) {
+    typeId = id;
```
This needs to throw if the type is not a decimal type, otherwise the caller can create an instance of something like `DType.INT32` and now it's no longer a singleton, breaking `==` and `!=` checks on it.  The documentation for this constructor should be updated to reflect the behavior.

jlowe replied:
This wasn't a private constructor when I first raised the issue.  😉 

Now that this is moved to being callable only from a factory and the factory has the checks, I'm happy.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+
+  public String getSimpleName() { return typeId.simpleName; }
+
+  // TODO: override hashCode
```
IDEs auto-generate these things, no reason to leave this as a TODO.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+  // TODO: override hashCode
+  @Override
+  public boolean equals(Object o) {
+    DType other = (DType) o;
```
This needs to check if the other object is an instance of `DType` rather than assume.  IDEs can often generate the code for this method to handle it properly.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -339,44 +340,44 @@ long getEndListOffset(long index) {
    */
   public final float getFloat(long index) {
-    assert type == DType.FLOAT32 : type + " is not a supported float type.";
+    assert type == DType.FLOAT32 : type.typeId + " is not a supported float type.";
```
.equals

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -339,44 +340,44 @@ long getEndListOffset(long index) {
    */
   public final double getDouble(long index) {
-    assert type == DType.FLOAT64 : type + " is not a supported double type.";
+    assert type == DType.FLOAT64 : type.typeId + " is not a supported double type.";
```
.equals

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -339,44 +340,44 @@ long getEndListOffset(long index) {
    */
   public final boolean getBoolean(long index) {
-    assert type == DType.BOOL8 : type + " is not a supported boolean type.";
+    assert type == DType.BOOL8 : type.typeId + " is not a supported boolean type.";
```
.equals

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -339,44 +340,44 @@ long getEndListOffset(long index) {
    */
   public byte[] getUTF8(long index) {
-    assert type == DType.STRING : type + " is not a supported string type.";
+    assert type == DType.STRING : type.typeId + " is not a supported string type.";
```
.equals

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -403,9 +404,9 @@ public String getJavaString(long index) {
    */
   public byte[] getBytesFromList(long rowIndex) {
-    assert type == DType.LIST : type + " is not a supported list of bytes type.";
+    assert type == DType.LIST : type.typeId + " is not a supported list of bytes type.";
```
.equals

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -403,9 +404,9 @@ public String getJavaString(long index) {
+    assert type == DType.LIST : type.typeId + " is not a supported list of bytes type.";
     HostColumnVectorCore listData = children.get(0);
-    assert listData.type == DType.INT8 || listData.type == DType.UINT8  : type +
+    assert listData.type == DType.INT8 || listData.type == DType.UINT8  : type.typeId +
```
.equals

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -518,7 +519,7 @@ private Object readValue(int rowIndex) {
       case BOOL8: return offHeap.data.getBoolean(rowOffset);
       case STRING: return getString(rowIndex);
-      default: throw new UnsupportedOperationException("Do not support " + type);
+      default: throw new UnsupportedOperationException("Do not support " + type.typeId);
```
We should have a .toString on DType.

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -518,7 +519,7 @@ private Object readValue(int rowIndex) {
       case INT16: return offHeap.data.getShort(rowOffset);
       case BOOL8: return offHeap.data.getBoolean(rowOffset);
       case STRING: return getString(rowIndex);
-      default: throw new UnsupportedOperationException("Do not support " + type);
```
We should at least have a follow on issue to update this API to return something for the various Decimal types.

revans2 commented on java/src/main/java/ai/rapids/cudf/JCudfSerialization.java:
```diff
@@ -200,7 +200,7 @@ public void writeTo(DataWriter dout) throws IOException {
       for (int i = 0; i < numColumns; i++) {
-        dout.writeInt(types[i].nativeId);
+        dout.writeInt(types[i].getNativeId());
         dout.writeInt((int) nullCounts[i]);
```
We need to update the code here to deal with scale too.

nartal1 replied:
done

revans2 commented on java/src/main/java/ai/rapids/cudf/Scalar.java:
```diff
@@ -253,7 +253,7 @@ public static Scalar durationFromLong(DType type, long value) {
         }
         return durationDaysFromInt(intValue);
       } else {
```
We at least need a follow on issue to support decimal scalar values.

nartal1 replied:
Scalar is mentioned as one of the tasks in umbrella issue https://github.com/rapidsai/cudf/issues/6515. Please let me know if that is okay to track scalar there.

revans2 replied:
That is fine. I mostly want to be sure that we don't forget it.

revans2 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+    STRING(0, 23, "str"),
+    LIST(0, 24, "list"),
+    DECIMAL32(4, 25, "decimal32"),
+    DECIMAL64(8, 26, "decimal64"),
```
decimal32 and decimal64 are not supported by CSV for parsing so we should not set a string indicating that they are. I have filed a follow on issue because this is a bug in other places in this code too.  We really should modify DType to just not have the strings here, and have the mapping closer to where it really matters.

https://github.com/rapidsai/cudf/issues/6574

sperlingxx replied:
Hi @revans2 , is it mean that we need to modify method `csv_reader_options::builder.dtypes` to pass multiple arguments rather than a single string for data type representation?

razajafri replied:
@revans2 please correct me if I am wrong but I think what you are saying is that we should leave the string blank or indicate in the string that its not supported yet. 

e.g. `DECIMAL32(4, 25, "NOT SUPPORTED")` 

Moving the string to a schema may be out of scope for this PR but that's a good point. Should we create a follow-on for that @revans2?

revans2 commented on java/src/main/java/ai/rapids/cudf/Table.java:
```diff
@@ -2333,7 +2333,7 @@ public TestBuilder timestampSecondsColumn(Long... values) {
 
     private static ColumnVector from(DType type, Object dataArray) {
```
When will these APIs be updated to support decimal types?  Had to test if we cannot create a ColumnVector from the data.

nartal1 replied:
My initial plan was to put up a small PR to make modifications to DType so that we could handle `scale`  and make sure we don't break any existing tests.  So I had not made any changes in the JNI to support decimal type. I think adding minimal decimal type support(non scalar) and adding corresponding tests makes sense as you said before.

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -286,7 +286,7 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_ifElseSS(JNIEnv *env, j
                                                                 jlong j_col_view,
                                                                 jlong j_agg,
-                                                                jint j_dtype) {
+                                                                jint j_dtype, jint scale) {
```
We need to use the scale not just pass it in.

nartal1 replied:
Had thought of doing this in follow on PR. Will make the modifications here itself.

nartal1 replied:
done

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -694,7 +694,8 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_dayOfYear(JNIEnv *env,
 JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_castTo(JNIEnv *env, jobject j_object,
-                                                                jlong handle, jint type) {
+                                                                jlong handle, jint type,
+                                                                jint scale) {
```
We need to use the scale not just pass it in.

nartal1 replied:
done

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -996,7 +997,8 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_stringConcatenation(
                                                                     jlong lhs_view, jlong rhs_view,
-                                                                    jint int_op, jint out_dtype) {
+                                                                    jint int_op, jint out_dtype,
+                                                                    jint scale) {
```
We need to use the scale not just pass it in.

nartal1 replied:
done

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1014,7 +1016,8 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_binaryOpVV(JNIEnv *env,
                                                                     jlong lhs_view, jlong rhs_ptr,
-                                                                    jint int_op, jint out_dtype) {
+                                                                    jint int_op, jint out_dtype,
+                                                                    jint scale) {
```
We need to use the scale not just pass it in.

nartal1 replied:
done

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1297,15 +1300,20 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_hash(JNIEnv *env,
     cudf::type_id n_type = static_cast<cudf::type_id>(j_type);
-    cudf::data_type n_data_type(n_type);
+    cudf::data_type n_data_type;
+    if ( n_type == cudf::type_id::DECIMAL32 || n_type == cudf::type_id::DECIMAL64 ) {
```
We should have a helper function that will take j_type and j_scale and return a data_type, so we can use it everywhere.

nartal1 replied:
created helper function which does the above.

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1619,13 +1639,19 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeColumnView(JNI
     cudf::type_id n_type = static_cast<cudf::type_id>(j_type);
-    cudf::data_type n_data_type(n_type);
+    cudf::data_type n_data_type;
+    if ( n_type == cudf::type_id::DECIMAL32 || n_type == cudf::type_id::DECIMAL64 ) {
```
This should be a helper function that gets used everywhere.

sperlingxx replied:
I've an idea that we pass `jobject dtype` instead of type_id and scale. And we create a helper function to convert `jobject dtype` to `cudf::data_type` via JNI. SetIntField. We can also build helper function for transaction from  `cudf::data_type`  to `jobject dtype`  in the same way.

jlowe replied:
> I have an idea that we pass jobject dtype instead of type_id and scale. And we create a helper function who converts jobject dtype to cudf::data_type via JNI.

Please _do not do this_.  Having the JNI call back into the JVM is quite expensive in practice.  It will be much faster to pass the elements of an object to the JNI call rather than an object pointer to JNI and require the JNI function to call back into the JVM to access the instance's fields.

sperlingxx replied:
@jlowe Thanks for pointing out that. I didn't realize JVM callback in the JNI env is quite inefficient.

revans2 commented on java/src/main/native/src/ScalarJni.cpp:
```diff
@@ -395,7 +395,7 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_Scalar_makeTimestampTimeScalar(JNIEn
 JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_Scalar_binaryOpSV(JNIEnv *env, jclass, jlong lhs_ptr,
                                                               jlong rhs_view, jint int_op,
-                                                              jint out_dtype) {
+                                                              jint out_dtype, jint scale) {
```
We need to use the scale not just pass it in.

revans2 commented on java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:
```diff
@@ -703,7 +703,11 @@ void testSequenceOtherTypes() {
+    for (DType.DTypeEnum type : DType.DTypeEnum.values()) {
+      // Decimal type not supported yet. Update this once it is supported.
+      if (type == DType.DTypeEnum.DECIMAL32 || type == DType.DTypeEnum.DECIMAL64) {
+        continue;
```
Is this going to be a follow on issue?

nartal1 replied:
I was hoping this could be covered in this https://github.com/rapidsai/cudf/issues/6515 as well. But if you think these things should not be covered in umbrella bug, I will create another issue. Please let me know.

revans2 replied:
OK

revans2 commented on java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:
```diff
@@ -747,7 +751,7 @@ void testFromScalarZeroRows() {
         case TIMESTAMP_MICROSECONDS:
         case TIMESTAMP_NANOSECONDS:
-          s = Scalar.timestampFromLong(type, 1234567890123456789L);
+          s = Scalar.timestampFromLong(new DType(type, 0), 1234567890123456789L);
```
This is why we want a factory because now the test is creating a new DType class when it does not need to.

nartal1 replied:
done

revans2 commented on java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:
```diff
@@ -770,7 +774,7 @@ void testFromScalarZeroRows() {
 
         try (ColumnVector c = ColumnVector.fromScalar(s, 0)) {
-          assertEquals(type, c.getType());
+          assertEquals(type, c.getType().typeId);
```
Why not do all of this in terms of DType instead of DTypeEnum so we get the scale check in here too.

nartal1 replied:
Sorry, there wasn't easy way to get all the types for test here. Will try to do in in terms of DType.

nartal1 replied:
This is resolved by @sperlingxx's PR which is merged to this PR.

revans2 reviewed (commented):
Also none of the tests have been updated to see if we can actually process decimals.  I don't feel good trying to merge in code that we are not testing.

nartal1 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+   * @param decimalScale scale - number of digits to the right of the decimal point in a number
    */
-  TIMESTAMP_SECONDS(8, 13, "timestamp[s]"),
+  public DType(DTypeEnum id, int decimalScale) {
```
Based on the comment here : https://github.com/rapidsai/cudf/pull/6514#discussion_r509303684 
There will  be one factory which takes both native DType  and scale, correct? There can't be one which doesn't scale.
Sorry if I am missing something here.

nartal1 replied:
Please ignore above comment.  The factory you mentioned should take an enum and the ones which I was pointing to was taking native DType and scale.

nartal1 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+  }
+
+  public static DType fromNative(int nativeId) {
+    return fromNative(nativeId, Integer.MIN_VALUE);
```
Sorry, does this mean we have only one factory which takes both native DType and scale? i.e scale is `0` for non-decimal types. 
The approach should be:
If it's non decimal, return singleton and for decimal - create new object.
Below code should be improved but wanted to know if my understanding 

``` 
public static DType fromNative(int nativeId, int scale) {
    if (nativeId >=0 && nativeId < staticDTypes.length) {
      if (staticDTypes[nativeId] != null) {
        return staticDTypes[nativeId];
      }
      if (nativeId == DTypeEnum.DECIMAL32.nativeId) {
        return new DType(DTypeEnum.DECIMAL32, scale);
      }
      if (nativeId == DTypeEnum.DECIMAL64.nativeId) {
        return new DType(DTypeEnum.DECIMAL64, scale);
      }
    }
    throw new IllegalArgumentException("Could not translate " + nativeId + " into a DType");
  }
```

nartal1 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+  // TODO: override hashCode
+  @Override
+  public boolean equals(Object o) {
+    DType other = (DType) o;
```
Thanks! I updated  `.equals()` and `.hashCode()` with auto-generated by IDE as per @jlowe 's suggestion. I didn't know that before.

nartal1 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -3270,7 +3273,15 @@ private void setNativeNullCount(int nullCount) throws CudfException {
+      if (dType.isBackedByDecimal()) {
+        return DType.fromNative(dType.getNativeId(), getNativeTypeScale(getViewHandle()));
+      }
+      return dType;
```
done

nartal1 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -3499,7 +3510,7 @@ private long getViewHandle() {
       long validAddr = valid == null ? 0 : valid.address;
       int nc = nullCount.orElse(OffHeapState.UNKNOWN_NULL_COUNT).intValue();
-      return makeCudfColumnView(dataType.nativeId, dataAddr, dataLen, offsetAddr, validAddr, nc,
+      return makeCudfColumnView(dataType.typeId.nativeId, 0 , dataAddr, dataLen, offsetAddr, validAddr, nc,
```
My bad. Updated it.

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -38,10 +38,12 @@
 
   protected final OffHeapState offHeap;
   protected final DType type;
+  protected long scale;
```
done

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -309,7 +310,7 @@ long getStartStringOffset(long index) {
    */
   long getStartListOffset(long index) {
-    assert type == DType.STRING || type == DType.LIST: type +
+    assert type == DType.STRING || type == DType.LIST: type.typeId +
```
done

nartal1 commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -327,7 +328,7 @@ long getEndStringOffset(long index) {
    */
   long getEndListOffset(long index) {
-    assert type == DType.STRING || type == DType.LIST: type +
+    assert type == DType.STRING || type == DType.LIST: type.typeId +
```
done

nartal1 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -17,76 +17,202 @@
+   * @return
    */
-  TIMESTAMP_MICROSECONDS(8, 15, "timestamp[us]"),
+  public int getNativeId() { return typeId.nativeId; }
```
Moved getNativeId to DataTypeEnum. Please let me know if it looks okay now.

revans2 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -116,88 +268,111 @@ public boolean isBackedByInt() {
+   *       DType.DECIMAL32,
+   *       DType.DECIMAL64
+   */
+  public boolean isBackedByDecimal() { return DECIMALS.contains(this.typeId); }
```
to me `isBackedBy` indicates how the data is stored. I would rather see this be called `isDecimalType` to match `isDurationType` and `isNestedType`.

revans2 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -116,88 +268,111 @@ public boolean isBackedByInt() {
 
-  public int getNativeId() {
-    return nativeId;
+  public boolean isTimestamp() {
```
nit: can we update this to be `isTimestampType` so it is consistent with the other APIs?

jlowe replied:
Note that `isTimestamp` is a pre-existing `DType` method.  Maybe add `isTimestampType` and deprecate this method?

revans2 replied:
@jlowe thanks I missed that. I saw the comments other places asking for us to use isTimestamp and I assumed it was new.  Also this is a nit and if we keep it as it I am fine with it.

nartal1 replied:
Added `isTimestampType` and deprecated `isTimestamp`

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -84,6 +84,20 @@ cudf::data_type timestamp_to_duration(cudf::data_type dt) {
+  return n_type == cudf::type_id::DECIMAL32 || n_type == cudf::type_id::DECIMAL64 ;
+}
+
+cudf::data_type update_if_decimal_type(jint out_dtype, jint scale) {
```
I don't think this name is clear.  I would rather see it be something like `make_data_type`.

nartal1 replied:
Thanks for the suggestion. done

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -84,6 +84,20 @@ cudf::data_type timestamp_to_duration(cudf::data_type dt) {
   return cudf::data_type(duration_type_id);
 }
 
+bool decimal_type(cudf::type_id n_type) {
```
Can we call this `is_decimal_type`

nartal1 replied:
Done. Thanks for the suggestion.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -16,77 +16,229 @@
+
+  @Override
+  public String toString() {
+    return "" + typeId;
```
This should be updated to print the scale iff its a decimal type

nartal1 replied:
I am not sure if this is the good way to print for decimal types. Any suggestions are welcome.

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -3270,7 +3273,11 @@ private void setNativeNullCount(int nullCount) throws CudfException {
+      return DType.fromNative(getNativeTypeId(getViewHandle()), getNativeTypeScale(getViewHandle()));
+    }
+
+    public int getNativeScale() {
```
Do we need this method, or at least does it need to be public?  Maybe I'm missing the use case where someone wants to access this without also calling `getNativeType` which provides the scale via the `DType` result.

nartal1 replied:
I had added this accessor method earlier but now i don't see any use case for it. i.e no reason to just get the scale. Will remove it.

revans2 replied:
It should not matter because OffHeapState, which this is a part of, is protected.  All the other methods are public too, but I am also fine with changing it.

jlowe commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -694,12 +709,13 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_dayOfYear(JNIEnv *env,
 }
 
 JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_castTo(JNIEnv *env, jobject j_object,
```
```suggestion
JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_castTo(JNIEnv *env, jclass,
```

nartal1 replied:
done

jlowe commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1297,15 +1317,15 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_hash(JNIEnv *env,
 
 JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
-    JNIEnv *env, jobject j_object, jint j_type, jlong j_data, jlong j_data_size, jlong j_offset,
+    JNIEnv *env, jobject j_object, jint j_type, jint scale, jlong j_data, jlong j_data_size, jlong j_offset,
```
```suggestion
    JNIEnv *env, jclass, jint j_type, jint scale, jlong j_data, jlong j_data_size, jlong j_offset,
```

nartal1 replied:
done

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -16,77 +16,229 @@
-   * Days since the UNIX epoch
+   * Constructor for Decimal Type
+   * @param id Enum representing data type.
+   * @param decimalScale scale - number of digits to the right of the decimal point in a number
```
Is this scale documentation correct?  If it's passed verbatim to libcudf then this is an exponent for the underlying value to reconstruct the decimal value being represented.

razajafri commented:
@revans2 @jlowe Depending on when we plan to officially support DecimalTypes fully, does it make sense to add a test that calls the public API's with a decimal type and makes sure it throws an exception so we know we are covered?

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -16,77 +16,229 @@
+    public int getNativeId() { return nativeId; }
+  }
+
+  final DTypeEnum typeId;
```
I don't see a public accessor for this.  Code outside of cudf will want to perform switch statements on data type IDs.

nartal1 replied:
done

revans2 commented:
> @revans2 @jlowe Depending on when we plan to officially support DecimalTypes fully, does it make sense to add a test that calls the public API's with a decimal type and makes sure it throws an exception so we know we are covered?

I think we need some kind of tests for this to go in.  We are partially adding in decimal type and it would be good to verify that if someone tries to use it that we get what we expect.  The hard part right now is that none of the code to produce decimals is written, so it might be difficult to test it if there is no way to make one

nartal1 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -1982,7 +1982,7 @@ public ColumnVector asTimestamp(DType timestampType, String format) {
     assert timestampType.isTimestamp() : "unsupported conversion to non-timestamp DType";
     return new ColumnVector(stringTimestampToTimestamp(getNativeView(),
-        timestampType.nativeId, format));
+        timestampType.typeId.getNativeId(), format));
```
done

nartal1 commented:
There are few pending items which I intend to address tomorrow.

revans2 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -20,6 +20,11 @@
+  below sections are updated as well:
+  1. Create a singleton object of the new type.
+  2. Update SINGLETON_DTYPE_LOOKUP to reflect new type. The order should be maintained between
+  DTypeEnum and SINGLETON_DTYPE_LOOKUP */
```
Looks good to me.

revans2 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -15,78 +15,292 @@
+  @Override
+  public String toString() {
+    if (isDecimalType()) {
+      return "" + typeId + " scale:" + scale;
```
nit: You don't need `"" +` in this case

revans2 commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -15,78 +15,292 @@
+    if (isDecimalType()) {
+      return "" + typeId + " scale:" + scale;
+    } else {
+      return "" + typeId;
```
nit: I find `typeId.toString` or `String.valueOf(typeId)` to be cleaner, even if it is larger.

revans2 reviewed (commented):
Looking good just a few more nits for me.

nartal1 commented:
Thanks to @sperlingxx for adding scalar tests. I have merged it in this branch. Sorry for bringing up this late. Thought of adding this  here to have some tests.

revans2 commented on java/src/main/native/src/dtype_utils.hpp:
```diff
@@ -0,0 +1,60 @@
+ */
+#pragma once
+
+namespace dtype_utils{
```
nit: This is really minor because all of the functions are inline and the header is private to the JNI code, but I would prefer to have the namespace have something with cudf and java/jni in it.

nartal1 replied:
Thanks @mythrocks , I have made the changes as per suggestion.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -15,78 +15,292 @@
-  public boolean isTimestamp() {
-    return TIMESTAMPS.contains(this);
+  /**
+   * Factory method specialized for decimal DType instances.
```
This public documentation should explain what `scale` means since many will find it unintuitive.

nartal1 replied:
Thanks! I have added comments. Please let me know if this is better.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -15,78 +15,292 @@
+
   /**
-   * ms since the UNIX epoch
+   * Returns scale for Decimal Type
```
This should explain what scale is since many will find it unintuitive compared to Java's `BigDecimal`.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -15,78 +15,292 @@
+  /**
+   * Factory method for DType instances
+   * @param nativeId nativeId of DataTypeEnun
+   * @param scale  scale should be provided for decimal type
```
This needs to elaborate on how `scale` is interpreted as many will find it unintuitive.

mythrocks commented on java/src/main/native/src/ScalarJni.cpp:
```diff
@@ -15,9 +15,11 @@
 #include <cudf/scalar/scalar_factories.hpp>
 
 #include "cudf_jni_apis.hpp"
+#include "dtype_utils.hpp"
```
Please note that `dtype_utils.hpp` is not independently usable in its current form. You can verify this by making it the first `#include` in `ScalarJni.hpp`. It works now because it is included after the other `cudf` headers.
Please `#include <cudf/types.hpp>` and the headers for any other symbols you need in `dtype_utils.hpp`.

mythrocks replied:
Thanks for incorporating this suggestion.

mythrocks reviewed (changes requested):
Some minor changes suggested.

nartal1 commented:
@jlowe @mythrocks  I think I have addressed review comments. Please take another look.
@razajafri Tagging you as well since you had few comments initially which were resolved :)

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -182,7 +182,9 @@ private DType(DTypeEnum id, int decimalScale) {
   /**
    * Returns scale for Decimal Type
-   * @return scale
+   * @return scale If negative, the scale is the number of digits to the right of
```
This comment implies that there's some special behavior if the scale is negative vs. >= 0 which is not the case.  It's always the base-10 exponent to multiply the unscaled value to produce the decimal value.  I also think the examples are great and should appear in all the methods that document what the scale parameter does.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -15,78 +15,302 @@
+  /**
+   * Factory method specialized for decimal DType instances.
+   * @param dt  enum corresponding to datatype.
+   * @return scale base-10 exponent to multiply the unscaled value to produce the decimal value.
```
```suggestion
   * @param scale base-10 exponent to multiply the unscaled value to produce the decimal value.
```

nartal1 replied:
Ah! I'm sorry. copy paste mistake.

jlowe commented on java/src/main/java/ai/rapids/cudf/DType.java:
```diff
@@ -15,78 +15,302 @@
+  /**
+   * Factory method for DType instances
+   * @param nativeId nativeId of DataTypeEnun
+   * @return scale base-10 exponent to multiply the unscaled value to produce the decimal value
```
```suggestion
   * @param scale base-10 exponent to multiply the unscaled value to produce the decimal value
```

nartal1 commented:
Thanks for the approvals!
